File: src/TelemetryService.cpp
#include "TelemetryService.h"

#include <WiFi.h>
#include <ArduinoJson.h>
#include <AsyncJson.h>
#include "Connector_Task.h"

TelemetryService::TelemetryService(AsyncWebServer* server) {
  server->on("/api/telemetry", HTTP_GET, [this](AsyncWebServerRequest* request) {
    handleTelemetry(request);
  });
}

void TelemetryService::handleTelemetry(AsyncWebServerRequest* request) {
  AsyncJsonResponse* response = new AsyncJsonResponse(false, 512);
  JsonObject root = response->getRoot();

  root["uptime_ms"] = millis();
  root["free_heap"] = ESP.getFreeHeap();
  root["min_free_heap"] = ESP.getMinFreeHeap();
  root["max_alloc_heap"] = ESP.getMaxAllocHeap();
  root["wifi_connected"] = WiFi.status() == WL_CONNECTED;
  if (WiFi.status() == WL_CONNECTED) {
    root["wifi_rssi"] = WiFi.RSSI();
  }

  JsonObject queues = root.createNestedObject("queues");
  if (gHybridConnectionQueue) {
    UBaseType_t spaces = uxQueueSpacesAvailable(gHybridConnectionQueue);
    queues["hybrid_available"] = spaces;
  }
  if (gCloudEventQueue) {
    UBaseType_t spaces = uxQueueSpacesAvailable(gCloudEventQueue);
    queues["cloud_available"] = spaces;
  }

  response->setLength();
  request->send(response);
}

File: fissioninterface/src/api/TelemetryService.ts
import axiosInstance from './axiosInstance';
import { API_CONFIG } from './apiConfig';

export interface TelemetryResponse {
  uptime_ms: number;
  free_heap: number;
  min_free_heap: number;
  max_alloc_heap: number;
  wifi_connected: boolean;
  wifi_rssi?: number;
  queues?: {
    hybrid_available?: number;
    cloud_available?: number;
  };
  [key: string]: any;
}

class TelemetryService {
  async fetchTelemetry(): Promise<TelemetryResponse> {
    const response = await axiosInstance.get(API_CONFIG.TELEMETRY.STATUS);
    return response.data as TelemetryResponse;
  }
}

const telemetryService = new TelemetryService();

export default telemetryService;
